### PR TITLE
Add ownership getters + setters for Array-type coefficients

### PR DESCRIPTION
**Pull request by j-signorelli**
This PR adds `GetOwnership()` and `SetOwnership()` methods for `VectorArrayCoefficient`, `MatrixArrayCoefficient`, and `MatrixArrayVectorCoefficient`, as there is currently no straightforward way to take/give ownership of a coefficient to objects of these types after calling `Set()`

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#4469](https://github.com/mfem/mfem/pull/4469) | @j-signorelli | @tzanio | @termi-official + @mlstowell | 8/25/24 | 11/3/24 | 11/10/24 | |

<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

### CODE REVIEW

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

**Comment by j-signorelli** [comment]
Hi @termi-official,

Thanks for the feedback, good points. Here are my thoughts: beyond a use-case for my own codebase I'm writing with MFEM where I want to take ownership dynamically (I can elaborate on this if you think it'd be useful to know), I feel that it's better to give MFEM users more functionality / capability to manage the memory on their own _if they choose to_ rather than limit the functionality under the premise that the users will be unable to do so.

Additionally but maybe more convincing is that the `ownCoeff` arrays are all private (as it seems most member variables of coefficients are set to be), so any derived classes of these types don't have the ability to know nor change ownership after the fact - this PR allows this functionality too.

Punchline is that, beyond wanting the added control over the memory for my own case, I think just in general allowing users to have more control over the memory of these objects and of derived types would be a nice feature.

**Comment by j-signorelli** [comment]
Just want to ping @termi-official @mlstowell about this as a reminder -- curious to hear thoughts and continue the discussion on whether this is a worthwhile PR or not. Right now I have just manually copied `VectorArrayCoefficient` into my own code with this above implementation.

Thanks!

**Review by mlstowell — APPROVED** [review]
I appreciate the flexibility these types of get/set methods provide. Of course these could lead to bugs but then the initial ownership setting could lead to similar bugs. I'm not very concerned about such issues because developers always need to be aware of potential memory management issues. Also coefficient objects are typically very light-weight so the potential hazard is minimal.

**Comment by tzanio** [comment]
ping: @termi-official

**Comment by termi-official** [comment]
Will come back to this next week!

File: fem/coefficient.hpp [hunk 1]
```diff
@@ -798,6 +798,12 @@ class VectorArrayCoefficient : public VectorCoefficient
    /// Sets coefficient in the vector.
    void Set(int i, Coefficient *c, bool own=true);
 
+   /// Set ownership of the i'th coefficient
+   void SetOwnership(int i, bool own) { ownCoeff[i] = own; };
```
Comment (termi-official) [inline fem/coefficient.hpp:802]:
```suggestion
   void SetOwnership(int i, bool own) { ownCoeff[i] = own; }
```

**Review by termi-official — APPROVED** [review]
Since @mlstowell does not share my raised concern I do not want to be overly critical here and will also give green light on my end. However, it might be worth a few thoughts on introducing smart pointers, starting from MFEM 5.0, for the ownership management, as it takes away some of the mental overhead for downstream libraries and applications.

I also found some minor style issue. We do not need the trailing `;` for method definitions in headers.

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Pull request merged**